rust/src/pgsql/parser.rs

```diff
@@ -0,0 +1,105 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+ use nom::Err;
+ use nom::error::ErrorKind;
+ use nom::IResult;
+
+pub const PGSQL_LENGTH_SIZE: usize = 4;
+pub const PGSQL_PROTO_SIZE: usize = 4;
+
+// PostgreSQL Startup message packet
+#[derive(Debug, PartialEq)]
+pub struct PGSQLStartupPacket<'a> {
+    pub length: u32,
+    pub proto_version_major: u16,
+    pub proto_version_minor: u16,
+    pub data: &'a[u8],
+}
+
+impl<'a> PGSQLStartupPacket<'a> {
+    pub fn is_valid(&self) -> bool {
+        let valid: bool = match self.proto_version_major {
+            1 | 2 | 3 => true,
+            _ => false,
+        };
+        valid
+    }
+}
+
+named!(pub parse_pgsql_startup_packet<PGSQLStartupPacket>,
+    do_parse!(
+        len: bits!(take_bits!(32u32))
+        >> proto_version: bits!(tuple!(
+            take_bits!(16u16),
+            take_bits!(16u16)))
+        >> data: take!(len  as usize - PGSQL_LENGTH_SIZE - PGSQL_PROTO_SIZE)
+        >> (PGSQLStartupPacket {
+            length: len,
+            proto_version_major: proto_version.0,
+            proto_version_minor: proto_version.1,
+            data: data,
+        })
+));
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_parse_pgsql_startup_packet() {
```

Comment:
I think a good next step would be to add a invalid input. For example one where the length field contains a value smaller than `PGSQL_LENGTH_SIZE + PGSQL_PROTO_SIZE`